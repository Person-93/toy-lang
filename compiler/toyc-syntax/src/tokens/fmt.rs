use crate::tokens::{NumLitPrefix, NumLitType};
use std::fmt::{Debug, Display, Formatter};

impl Display for super::Ident {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", self.0)
  }
}

impl Debug for super::Ident {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "Ident({self})")
  }
}

impl Display for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    Display::fmt(&self.0, f)
  }
}

impl Debug for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "StrLit({self})")
  }
}

impl Display for super::NumLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    let super::NumLit {
      prefix,
      val,
      decimal,
      ty,
    } = self;

    let prefix = match prefix {
      None => "",
      Some(NumLitPrefix::Hex) => "0x",
      Some(NumLitPrefix::Binary) => "0b",
    };

    let decimal = match decimal {
      None => String::new(),
      Some(n) => format!("{n}"),
    };

    let ty = match ty {
      None => "",
      Some(NumLitType::Float) => "f",
      Some(NumLitType::Int) => "i",
      Some(NumLitType::Unsigned) => "u",
    };

    write!(f, "{prefix}{val}{decimal}{ty}")
  }
}
