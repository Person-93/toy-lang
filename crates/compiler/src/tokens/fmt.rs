use crate::tokens::{NumLitPrefix, NumLitType};
use std::fmt::{Display, Formatter};

impl Display for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "\"{}\"", self.0)
  }
}

impl Display for super::BoolLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", if self.0 { "true" } else { "false" })
  }
}

impl Display for super::NumLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    let super::NumLit {
      prefix,
      val,
      decimal,
      ty,
    } = self;

    let prefix = match prefix {
      None => "",
      Some(NumLitPrefix::Hex) => "0x",
      Some(NumLitPrefix::Binary) => "0b",
    };

    let decimal = match decimal {
      None => String::new(),
      Some(n) => format!("{n}"),
    };

    let ty = match ty {
      None => "",
      Some(NumLitType::Float) => "f",
      Some(NumLitType::Int) => "i",
      Some(NumLitType::Unsigned) => "u",
    };

    write!(f, "{prefix}{val}{decimal}{ty}")
  }
}
