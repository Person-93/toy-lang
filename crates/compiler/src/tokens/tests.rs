use super::*;
use std::fmt::{Debug, Formatter};

use insta::{assert_snapshot, glob};
use logos::{Logos, Span};
use std::fs;

#[test]
fn snapshot() {
  glob!("toys/*.toy", |file| {
    let text = fs::read_to_string(file).unwrap();
    assert_snapshot!(Token::lexer(&text)
      .spanned()
      .map(|(token, span)| Output(token, span.clone(), &text[span]).to_string())
      .collect::<Vec<_>>()
      .join("\n"))
  });

  struct Output<'a>(Token, Span, &'a str);

  impl std::fmt::Display for Output<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
      if let Token::Error = self.0 {
        write!(f, "ERROR @ {:?}: \"{}\"", self.1, self.2)
      } else {
        write!(f, "{:?} @ {:?}", self.0, self.1)
      }
    }
  }
}
