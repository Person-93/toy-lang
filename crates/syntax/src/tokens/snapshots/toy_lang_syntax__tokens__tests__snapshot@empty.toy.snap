---
source: crates/compiler/src/tokens/tests.rs
assertion_line: 14
expression: "Token::lexer(&text).spanned().map(|(token, span)|\n                                      Output(token, span.clone(),\n                                             &text[span]).to_string()).collect::<Vec<_>>().join(\"\\n\")"
input_file: toys/empty.toy

---
